### PR TITLE
Improve error messages for missing @nogc inference

### PR DESCRIPTION
**Pull request by MoonlightSentinel**
Provide an explanaition why `@nogc` could not be inferred when calling a templated function that violates `@nogc`.

TODO:
- evaluate other implementations (especially w.r.t. implementing hints for other attributes)
  - See #12383
- more tests
- update C++ headers
- changelog?

### CODE REVIEW

File: src/dmd/func.d [hunk 1]
```diff
@@ -2540,6 +2540,7 @@ extern (C++) class FuncDeclaration : Declaration
         {
             tf = new TypeFunction(ParameterList(fparams), treturn, LINK.c, stc);
             fd = new FuncDeclaration(Loc.initial, Loc.initial, id, STC.static_, tf);
+            fd.generated = true;
```
Comment (TungstenHeart) [inline src/dmd/func.d:2543]: this change seems unrelated, can you explain ?

Comment (MoonlightSentinel) [reply src/dmd/func.d:2543]: This changes `fd.kind()` to `generated function` which is more informative when `@nogc` inference fails due to a Druntime lowering.

Not sure if I'll keep this around because the check now recognises `foreach`-lowerings.

**Comment by nordlow** [comment]
> Provide an explanaition why `@nogc` could not be inferred when calling a templated function that violates `@nogc`.

Wonderful that diagnostics for this is finally happening, thank you. Are you planning simiilar diagnostics for other qualifers?

**Comment by MoonlightSentinel** [comment]
I'll try. These diagnostics are possible for `@nogc` (and maybe `nothrow` as well) because those checks are implemented using a dedicated visitor. 

`pure` and `@safe` will probably be a lot harder.

**Comment by nordlow** [comment]
> I'll try. These diagnostics are possible for `@nogc` (and maybe `nothrow` as well) because those checks are implemented using a dedicated visitor.

Thank you.

**Comment by RazvanN7** [comment]
Please ping me when this is ready for review.

**Comment by thewilsonator** [comment]
also 
```
dmd/identifier.d(145:26)[warn]: Useless final attribute, functions declared within a final class are never virtual
```

**Comment by MoonlightSentinel** [comment]
I'll close this PR for now because #12383 looks more promising.
